libsmtutil/genz3wrapper.py

```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env python
+
+import sys
+import re
+
+# Patterns to match Z3 API entry point definitions.
+def_pat = re.compile(" *def_API(.*)")
+extradef_pat = re.compile(" *extra_API(.*)")
+# Pattern to extract name and arguments from the above.
+def_args_pat = re.compile("\('(.*)'[^\(\)]*\((.*)\)\s*\)")
+# Pattern to extract a list of arguments from the above.
+arg_list_pat = re.compile("[^\(]*\([^\)]*\)[, ]*")
+
+def generateEntryPoint(line, args):
+    m = def_args_pat.match(args)
+    if m:
+        name = m.group(1)
+        num_args = len(arg_list_pat.findall(m.group(2)))
+        arglist = ''
+        paramlist = ''
+        for i in range(num_args):
+            if i != 0:
+                arglist += ', '
+                paramlist += ', '
+            arglist += '_' + str(i)
+            paramlist += 'ArgType<&' + name + ', ' + str(i) + '> _' + str(i)
+        print('ResultType<&' + name + '> Z3_API ' + name + '(' + paramlist + ')')
+        print('{')
+        print('\t' + 'static auto sym = reinterpret_cast<decltype(&' + name + ')>(Z3Loader::get().loadSymbol(\"' + name + '\"));')
+        print('\treturn sym(' + arglist + ');')
+        print('}')
+    else:
+        raise Exception('Could not parse entry point definition: ' + line)
+
+print(r"""// This file is auto-generated from genwrapper.py
+#include <z3.h>
+#include <tuple>
+#include <libsmtutil/Z3Loader.h>
+
+namespace {
```

Comment:
I actually disagree that this makes the script unreadable in any way :-). On the contrary - if I used a template it would be less clear what I'll end up generating in contrast to just having it here. I don't think we should nor have to overengineer this.

Reply:
My problem with this is that it breaks syntax highlighting and that different indentation between the parts makes me confused about the structure of the file. I had to read it a few times when I first encountered it before I understood what's happening there.

And I don't think that it's overengineering. Loading and filling the template could be as short as `open('z3wrapper.h.template'm 'r').read().replace('<#PLACEHOLDER#>', generated_functions)`. 1 more line if you want to use the context manager to close the file properly.

As to the general structure of the script - I personally prefer to have scripts structured with proper functions for everything and data passed via parameters rather than globals. I don't think that putting stuff at global scope necessarily makes them simpler. I have easier time understanding them when I can focus on smaller, isolated bits that I know cannot be affected by the stuff I'm not looking at.